File: symfony/src/Service/Players/PlayerAchievementService.php
<?php

namespace App\Service\Players;

use App\Entity\Players\PlayerAchievement;
use App\Repository\Players\PlayerAchievementRepository;

class PlayerAchievementService
{
    public function __construct(
        private PlayerAchievementRepository $repository,
    ) {}

    public function create(array $data): PlayerAchievement
    {
        throw new \LogicException('Not implemented');
    }

    public function update(PlayerAchievement $entity, array $data): PlayerAchievement
    {
        throw new \LogicException('Not implemented');
    }

    public function incrementProgress(int $id, $amount): void
    {
        $entity = $this->repository->find($id);
        if (!$entity) throw new \RuntimeException('PlayerAchievement not found: ' . $id);
        $entity->incrementProgress($amount);
        $this->repository->save($entity, flush: true);
    }

    public function complete(int $id): void
    {
        $entity = $this->repository->find($id);
        if (!$entity) throw new \RuntimeException('PlayerAchievement not found: ' . $id);
        $entity->complete();
        $this->repository->save($entity, flush: true);
    }

    public function setIsCompleted(int $id, string $value): void
    {
        $entity = $this->repository->find($id);
        if (!$entity) throw new \RuntimeException('PlayerAchievement not found: ' . $id);
        $entity->setIsCompleted($value);
        if ($value === 'TRUE') {
            $entity->complete(); // @on(is_completed = true)
        }
        $this->repository->save($entity, flush: true);
    }
}

File: rails/app/services/players/player_achievement_service.rb
module Players
  class PlayerAchievementService
    def initialize(repository = PlayerAchievement)
      @repository = repository
    end

    def create(attributes)
      raise NotImplementedError, "#{self.class}#create not implemented"
    end

    def update(entity, attributes)
      raise NotImplementedError, "#{self.class}#update not implemented"
    end

    def increment_progress(id, amount)
      instance = PlayerAchievement.find(id)
      instance.increment_progress(amount)
      instance.save!
    end

    def complete(id)
      instance = PlayerAchievement.find(id)
      instance.complete()
      instance.save!
    end

    # triggered by @on(is_completed = true)
    def set_is_completed(id, value)
      instance = PlayerAchievement.find(id)
      instance.is_completed = value
      if value.to_s.upcase == 'TRUE'
        instance.complete
      end
      instance.save!
    end
  end
end
